### PR TITLE
Add check for 0 to 1 inclusive for elements of target tensor in BCE loss

### LINKED ISSUE #87373: Check on target being between 0-1 for binary cross entropy loss
### 📚 The doc issue

In the docstring  of `https://pytorch.org/docs/stable/_modules/torch/nn/functional.html#binary_cross_entropy` it is mentioned that `target: Tensor of the same shape as input with values between 0 and 1.` However there is no such check added anywhere. 

### Suggest a potential alternative/fix

Add a check similar to the one for the input [here](https://github.com/pytorch/pytorch/blob/d56017a14f34b5130fa70c0cba010e3d2506deb0/aten/src/ATen/native/Loss.cpp#L236-L252)

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

### PR DESCRIPTION
**Pull request by kiersten-stokes**
Fixes #87373

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### CODE REVIEW

**Review by bdhirsh — APPROVED** [review]
Thanks for the fix!

**Review by bdhirsh — CHANGES_REQUESTED** [review]
actually, I see several nn tests failing - it looks like some of our tests pass > 1 target values through `binary_cross_entropy`. cc @mikaylagawarecki ? (let me know if I should cc someone else).

**Comment by kiersten-stokes** [comment]
> actually, I see several nn tests failing - it looks like some of our tests pass > 1 target values through `binary_cross_entropy`. cc @mikaylagawarecki ? (let me know if I should cc someone else).

@bdhirsh thanks for the review! I was also wondering about those tests but haven't had time to circle back (and wasn't sure who to ping 🙂 ). Looking forward to hearing thoughts on how/if we want to handle those!

**Review by mikaylagawarecki — COMMENTED** [review]
Hey @kiersten-stokes, thanks for sending this fix!

The tests failures arise because of tests where `target` was erroneously not constrained in `[0, 1]`. For most of the target outputs, we ensure this using `torch.randn(...).gt(0).double()` but this was missed for these two tests [here](https://github.com/pytorch/pytorch/blob/47dca20d800310b03c24c1ccd5fa6b9c7cae034c/torch/testing/_internal/common_nn.py#L4867) and [here](https://github.com/pytorch/pytorch/blob/47dca20d800310b03c24c1ccd5fa6b9c7cae034c/torch/testing/_internal/common_nn.py#L229)). Fixing these should resolve the errors.

Additionally, we should also [add the check on CUDA](https://github.com/pytorch/pytorch/blob/d56017a14f34b5130fa70c0cba010e3d2506deb0/aten/src/ATen/native/cuda/Loss.cu#L92)

**Comment by kiersten-stokes** [comment]
> The tests failures arise because of tests where `target` was erroneously not constrained in `[0, 1]`. For most of the target outputs, we ensure this using `torch.randn(...).gt(0).double()` but this was missed for these two tests [here](https://github.com/pytorch/pytorch/blob/47dca20d800310b03c24c1ccd5fa6b9c7cae034c/torch/testing/_internal/common_nn.py#L4867) and [here](https://github.com/pytorch/pytorch/blob/47dca20d800310b03c24c1ccd5fa6b9c7cae034c/torch/testing/_internal/common_nn.py#L229)). Fixing these should resolve the errors.

@mikaylagawarecki I really appreciate you pointing me to the relevant tests! It seems embarrassingly simple now that I see it - I guess I'm still learning how to comb my way through the logs efficiently 🙈 

> Additionally, we should also [add the check on CUDA](https://github.com/pytorch/pytorch/blob/d56017a14f34b5130fa70c0cba010e3d2506deb0/aten/src/ATen/native/cuda/Loss.cu#L92)

Done! Will stand by for CI failures (apologies, I have no way to run CUDA-based tests at the moment) and circle back should there be any failures similar to the above!

**Comment by mikaylagawarecki** [comment]
@kiersten-stokes Of course, happy to help!

Looking at the logs, it looks like I might have missed another one where target is not correctly constrained: https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L8923

It's hard to tell if there are more because CI jobs terminate when a test fail :( But you could try running tests on CPU, `python test/test_nn.py`, you can also try to filter tests by substring with the `-k` option e.g. `python test/test_nn.py -k cross_entropy` but depending on the test naming, that might not be a catch-all.

Let me know if you have any questions about local testing!

**Review by mikaylagawarecki — APPROVED** [review]
Thanks!
